File: domain/src/main/java/com/beat/domain/booking/domain/Booking.java
package com.beat.domain.booking.domain;

import java.time.LocalDateTime;

import com.beat.domain.performance.domain.BankName;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Booking {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private int purchaseTicketCount;

	@Column(nullable = false)
	private String bookerName;

	@Column(nullable = false)
	private String bookerPhoneNumber;

	@Column(nullable = false)
	@Enumerated(EnumType.STRING)
	private BookingStatus bookingStatus = BookingStatus.CHECKING_PAYMENT;

	@Column(nullable = false)
	private LocalDateTime createdAt = LocalDateTime.now();

	@Column(nullable = true)
	private LocalDateTime cancellationDate;

	@Column(nullable = true)
	private String birthDate;

	@Column(nullable = true)
	private String password;

	@Enumerated(EnumType.STRING)
	@Column(nullable = true)
	private BankName bankName;

	@Column(nullable = true)
	private String accountNumber;

	@Column(nullable = true)
	private String accountHolder;

	@Column(name = "schedule_id", nullable = false)
	private Long scheduleId;

	@Column(name = "user_id", nullable = false)
	private Long userId;

	@Builder
	private Booking(int purchaseTicketCount, String bookerName, String bookerPhoneNumber, BookingStatus bookingStatus,
		String birthDate, String password, BankName bankName, String accountNumber, String accountHolder,
		Long scheduleId, Long userId) {
		this.purchaseTicketCount = purchaseTicketCount;
		this.bookerName = bookerName;
		this.bookerPhoneNumber = bookerPhoneNumber;
		this.bookingStatus = bookingStatus;
		this.birthDate = birthDate;
		this.password = password;
		this.bankName = bankName;
		this.accountNumber = accountNumber;
		this.accountHolder = accountHolder;
		this.scheduleId = scheduleId;
		this.userId = userId;
	}

	public static Booking create(int purchaseTicketCount, String bookerName, String bookerPhoneNumber,
		BookingStatus bookingStatus, String birthDate, String password,
		BankName bankName, String accountNumber, String accountHolder, Long scheduleId, Long userId) {
		return Booking.builder()
			.purchaseTicketCount(purchaseTicketCount)
			.bookerName(bookerName)
			.bookerPhoneNumber(bookerPhoneNumber)
			.bookingStatus(bookingStatus)
			.birthDate(birthDate)
			.password(password)
			.bankName(bankName)
			.accountNumber(accountNumber)
			.accountHolder(accountHolder)
			.scheduleId(scheduleId)
			.userId(userId)
			.build();
	}

	public void updateBookingStatus(BookingStatus bookingStatus) {
		this.bookingStatus = bookingStatus;
		if (bookingStatus == BookingStatus.BOOKING_CANCELLED || bookingStatus == BookingStatus.BOOKING_DELETED) {
			this.cancellationDate = LocalDateTime.now();
		}
	}

	public void updateRefundInfo(BankName bankName, String accountNumber, String accountHolder) {
		this.bankName = bankName;
		this.accountNumber = accountNumber;
		this.accountHolder = accountHolder;
		this.bookingStatus = BookingStatus.REFUND_REQUESTED;
	}
}

File: domain/src/main/kotlin/com/beat/domain/booking/domain/Booking.kt
package com.beat.domain.booking.domain

import com.beat.domain.booking.exception.BookingErrorCode
import com.beat.domain.performance.domain.BankName
import com.beat.domain.schedule.domain.Schedule
import com.beat.domain.user.domain.Users
import com.beat.global.support.exception.BadRequestException
import java.time.LocalDateTime

@ConsistentCopyVisibility
data class Booking private constructor(
    private val bookingId: Id?,
    private val purchaseTicketCount: Int,
    private val bookerName: String,
    private val bookerPhoneNumber: String,
    private val bookingStatus: BookingStatus,
    private val createdAt: LocalDateTime,
    private val cancellationDate: LocalDateTime?,
    private val birthDate: String?,
    private val password: String?,
    private val bankName: BankName?,
    private val accountNumber: String?,
    private val accountHolder: String?,
    private val linkedScheduleId: Schedule.Id,
    private val linkedUserId: Users.Id,
) {
    fun getId(): Long? = bookingId?.value

    fun getPurchaseTicketCount(): Int = purchaseTicketCount

    fun getBookerName(): String = bookerName

    fun getBookerPhoneNumber(): String = bookerPhoneNumber

    fun getBookingStatus(): BookingStatus = bookingStatus

    fun getCreatedAt(): LocalDateTime = createdAt

    fun getCancellationDate(): LocalDateTime? = cancellationDate

    fun getBirthDate(): String? = birthDate

    fun getPassword(): String? = password

    fun getBankName(): BankName? = bankName

    fun getAccountNumber(): String? = accountNumber

    fun getAccountHolder(): String? = accountHolder

    fun getScheduleId(): Long = linkedScheduleId.value

    fun getUserId(): Long = linkedUserId.value

    fun updateBookingStatus(bookingStatus: BookingStatus): Booking = copy(
        bookingStatus = bookingStatus,
        cancellationDate = if (isTerminalCancellationStatus(bookingStatus)) {
            cancellationDate ?: LocalDateTime.now()
        } else {
            cancellationDate
        },
    )

    fun updateRefundInfo(bankName: BankName?, accountNumber: String?, accountHolder: String?): Booking = copy(
        bankName = bankName,
        accountNumber = accountNumber,
        accountHolder = accountHolder,
        bookingStatus = BookingStatus.REFUND_REQUESTED,
    )

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            @JvmStatic
            fun from(value: Long): Id = Id(value)

            @JvmStatic
            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        @JvmStatic
        fun create(
            purchaseTicketCount: Int,
            bookerName: String,
            bookerPhoneNumber: String,
            birthDate: String?,
            password: String?,
            bankName: BankName?,
            accountNumber: String?,
            accountHolder: String?,
            scheduleId: Long?,
            userId: Long?,
        ): Booking {
            validatePurchaseTicketCount(purchaseTicketCount)
            requireNotNull(scheduleId) { "scheduleId must not be null" }
            requireNotNull(userId) { "userId must not be null" }

            return Booking(
                bookingId = null,
                purchaseTicketCount = purchaseTicketCount,
                bookerName = bookerName,
                bookerPhoneNumber = bookerPhoneNumber,
                bookingStatus = BookingStatus.CHECKING_PAYMENT,
                createdAt = LocalDateTime.now(),
                cancellationDate = null,
                birthDate = birthDate,
                password = password,
                bankName = bankName,
                accountNumber = accountNumber,
                accountHolder = accountHolder,
                linkedScheduleId = Schedule.Id.from(scheduleId),
                linkedUserId = Users.Id.from(userId),
            )
        }

        @JvmStatic
        fun rehydrate(
            id: Long?,
            purchaseTicketCount: Int,
            bookerName: String,
            bookerPhoneNumber: String,
            bookingStatus: BookingStatus,
            createdAt: LocalDateTime,
            cancellationDate: LocalDateTime?,
            birthDate: String?,
            password: String?,
            bankName: BankName?,
            accountNumber: String?,
            accountHolder: String?,
            scheduleId: Long?,
            userId: Long?,
        ): Booking {
            requireNotNull(scheduleId) { "scheduleId must not be null" }
            requireNotNull(userId) { "userId must not be null" }

            return Booking(
                bookingId = Id.fromNullable(id),
                purchaseTicketCount = purchaseTicketCount,
                bookerName = bookerName,
                bookerPhoneNumber = bookerPhoneNumber,
                bookingStatus = bookingStatus,
                createdAt = createdAt,
                cancellationDate = cancellationDate,
                birthDate = birthDate,
                password = password,
                bankName = bankName,
                accountNumber = accountNumber,
                accountHolder = accountHolder,
                linkedScheduleId = Schedule.Id.from(scheduleId),
                linkedUserId = Users.Id.from(userId),
            )
        }

        private fun validatePurchaseTicketCount(purchaseTicketCount: Int) {
            if (purchaseTicketCount <= 0) {
                throw BadRequestException(BookingErrorCode.INVALID_DATA_FORMAT)
            }
        }

        private fun isTerminalCancellationStatus(bookingStatus: BookingStatus): Boolean =
            bookingStatus == BookingStatus.BOOKING_CANCELLED || bookingStatus == BookingStatus.BOOKING_DELETED
    }
}
